RecoLocalCalo/EcalRecProducers/plugins/alpaka/TimeComputationKernels.h

```diff
@@ -0,0 +1,1162 @@
+                printf(
+                    "ch = %d ltx = %d tmax = %f tmaxerr = %f time1 = %f time2 = %f offset = %f rmin = %f rmax = "
+                    "%f\n",
+                    ch,
+                    ltx,
+                    tmax,
+                    tmaxerr,
+                    time1,
+                    time2,
+                    offset,
+                    rmin,
+                    rmax);
+#endif
+
+              ScalarType sumAf = 0;
+              ScalarType sumff = 0;
+              const int itmin = std::max(-1, static_cast<int>(std::floor(tmax - alphabeta)));
+              auto loffset = (static_cast<ScalarType>(itmin) - tmax) * invalphabeta;
+              // TODO: data dependence
+              for (int it = itmin + 1; it < nsamples; ++it) {
+                loffset += invalphabeta;
+                if (useless_sample_values[ch_start + it])
+                  continue;
+                auto const inverr2 = 1. / (sample_value_errors[ch_start + it] * sample_value_errors[ch_start + it]);
+                auto const term1 = 1. + loffset;
+                auto const f = (term1 > 1e-6) ? fast_expf(alpha * (fast_logf(term1) - loffset)) : 0;
+                sumAf += sample_values[ch_start + it] * (f * inverr2);
+                sumff += f * (f * inverr2);
+              }
+
+              auto const sumAA = sumAAsNullHypot[ch];
+              auto const sum0 = sum0sNullHypot[ch];
+              chi2 = sumAA;
+              // TODO: sum0 can not be 0 below, need to introduce the check upfront
+              if (sumff > 0) {
+                chi2 = sumAA - sumAf * (sumAf / sumff);
+              }
+              chi2 /= sum0;
+
+#ifdef DEBUG_TC_MAKERATIO
+              if (ch == 1 || ch == 0)
+                printf(
+                    "ch = %d ltx = %d sumAf = %f sumff = %f sumAA = %f sum0 = %d tmax = %f tmaxerr = %f chi2 = "
+                    "%f\n",
+                    ch,
+                    ltx,
+                    sumAf,
+                    sumff,
+                    sumAA,
+                    static_cast<int>(sum0),
+                    tmax,
+                    tmaxerr,
+                    chi2);
+#endif
+
+              if (chi2 > 0 && tmax > 0 && tmaxerr > 0)
+                internalCondForSkipping2 = false;
+              else
+                chi2 = std::numeric_limits<ScalarType>::max();
+            }
+          }
+
+          // store into smem
+          shr_chi2s[idx.local] = chi2;
+          shr_chi2[idx.local] = chi2;
+          shr_tmax[idx.local] = tmax;
+          shr_tmaxerr[idx.local] = tmaxerr;
+          shr_condForUselessSamples[idx.local] = condForUselessSamples;
+          shr_internalCondForSkipping1[idx.local] = internalCondForSkipping1;
+          shr_internalCondForSkipping2[idx.local] = internalCondForSkipping2;
+        }
+
+        alpaka::syncBlockThreads(acc);
+
+        // find min chi2 - quite crude for now
+        // TODO validate/check
```

Comment:
if this still an open TODO ?

Reply:
This also comes from the current CUDA implementation (as all other TODOs as well) and should maybe looked at at some point but perhaps not for this PR.